libqtile/utils.py

```diff
@@ -420,13 +424,37 @@ async def add_signal_receiver(
         "/org/freedesktop/DBus",
         "AddMatch",
         "s",
-        rule,
+        [rule],
     )
 
     # Check if message sent successfully
     if msg and msg.message_type == MessageType.METHOD_RETURN:
         bus.add_message_handler(callback)
+        dbus_match_calls.append((bus, rule))
         return True
 
     else:
         return False
+
+
+async def remove_signal_receiver(bus: MessageBus, rule: str) -> bool:
```

Comment:
Hmm, good question. Probably dbus-next type stubs in the `stubs` folder. Check out stubgen (`stubgen -m dbus_next`) which can generate them for you.

Reply:
Need to do some reading about stubs. This generated lots of mypy errors:
```
2022-05-02T20:39:18.2849235Z stubs/dbus_next/_private/constants.pyi:1: error: Module "_typeshed" has no attribute "Incomplete"
2022-05-02T20:39:18.5896114Z stubs/dbus_next/_private/marshaller.pyi:2: error: Module "_typeshed" has no attribute "Incomplete"
2022-05-02T20:39:18.5943587Z stubs/dbus_next/signature.pyi:3: error: Module "_typeshed" has no attribute "Incomplete"
2022-05-02T20:39:18.5944085Z stubs/dbus_next/message.pyi:7: error: Module "_typeshed" has no attribute "Incomplete"
2022-05-02T20:39:18.5944566Z stubs/dbus_next/errors.pyi:4: error: Module "_typeshed" has no attribute "Incomplete"
2022-05-02T20:39:18.6802682Z stubs/dbus_next/_private/unmarshaller.pyi:6: error: Module "_typeshed" has no attribute "Incomplete"
2022-05-02T20:39:18.6829495Z stubs/dbus_next/_private/address.pyi:3: error: Module "_typeshed" has no attribute "Incomplete"
2022-05-02T20:39:18.7056791Z stubs/dbus_next/_private/util.pyi:5: error: Name "Tuple" is not defined
2022-05-02T20:39:18.7057337Z stubs/dbus_next/_private/util.pyi:5: note: Did you forget to import it from "typing"? (Suggestion: "from typing import Tuple")
2022-05-02T20:39:18.7113578Z stubs/dbus_next/introspection.pyi:6: error: Module "_typeshed" has no attribute "Incomplete"
2022-05-02T20:39:18.7114473Z stubs/dbus_next/introspection.pyi:15: error: Self argument missing for a non-static method (or an invalid type for self)
2022-05-02T20:39:18.7115086Z stubs/dbus_next/introspection.pyi:33: error: Self argument missing for a non-static method (or an invalid type for self)
2022-05-02T20:39:19.0616446Z stubs/dbus_next/service.pyi:6: error: Module "_typeshed" has no attribute "Incomplete"
2022-05-02T20:39:19.0616975Z stubs/dbus_next/proxy_object.pyi:8: error: Module "_typeshed" has no attribute "Incomplete"
2022-05-02T20:39:19.0617455Z stubs/dbus_next/message_bus.pyi:12: error: Module "_typeshed" has no attribute "Incomplete"
2022-05-02T20:39:19.0618096Z stubs/dbus_next/glib/proxy_object.pyi:16: error: Return type "List[ProxyObject]" of "get_children" incompatible with return type "List[BaseProxyObject]" in supertype "BaseProxyObject"
2022-05-02T20:39:19.0618830Z stubs/dbus_next/aio/proxy_object.pyi:17: error: Return type "List[ProxyObject]" of "get_children" incompatible with return type "List[BaseProxyObject]" in supertype "BaseProxyObject"
2022-05-02T20:39:19.0619433Z stubs/dbus_next/glib/message_bus.pyi:9: error: Module "_typeshed" has no attribute "Incomplete"
2022-05-02T20:39:19.0620078Z stubs/dbus_next/glib/message_bus.pyi:53: error: Argument 3 of "get_proxy_object" is incompatible with supertype "BaseMessageBus"; supertype defines the argument type as "Union[Node, str, Element]"
2022-05-02T20:39:19.0620683Z stubs/dbus_next/glib/message_bus.pyi:53: note: This violates the Liskov substitution principle
2022-05-02T20:39:19.0621652Z stubs/dbus_next/glib/message_bus.pyi:53: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
2022-05-02T20:39:19.0622217Z stubs/dbus_next/aio/message_bus.pyi:10: error: Module "_typeshed" has no attribute "Incomplete"
2022-05-02T20:39:19.0622747Z stubs/dbus_next/aio/message_bus.pyi:33: error: Signature of "introspect" incompatible with supertype "BaseMessageBus"
2022-05-02T20:39:19.0623216Z stubs/dbus_next/aio/message_bus.pyi:33: note:      Superclass:
2022-05-02T20:39:19.0623944Z stubs/dbus_next/aio/message_bus.pyi:33: note:          def introspect(self, bus_name: str, path: str, callback: Callable[[Optional[Node], Optional[Exception]], None]) -> Any
2022-05-02T20:39:19.0624742Z stubs/dbus_next/aio/message_bus.pyi:33: note:      Subclass:
2022-05-02T20:39:19.0625421Z stubs/dbus_next/aio/message_bus.pyi:33: note:          def introspect(self, bus_name: str, path: str, timeout: float = ...) -> Coroutine[Any, Any, Node]
2022-05-02T20:39:19.0626001Z stubs/dbus_next/aio/message_bus.pyi:34: error: Signature of "request_name" incompatible with supertype "BaseMessageBus"
2022-05-02T20:39:19.0626443Z stubs/dbus_next/aio/message_bus.pyi:34: note:      Superclass:
2022-05-02T20:39:19.0627641Z stubs/dbus_next/aio/message_bus.pyi:34: note:          def request_name(self, name: str, flags: NameFlag = ..., callback: Optional[Callable[[Optional[RequestNameReply], Optional[Exception]], None]] = ...) -> Any
2022-05-02T20:39:19.0628211Z stubs/dbus_next/aio/message_bus.pyi:34: note:      Subclass:
2022-05-02T20:39:19.0628893Z stubs/dbus_next/aio/message_bus.pyi:34: note:          def request_name(self, name: str, flags: NameFlag = ...) -> Coroutine[Any, Any, RequestNameReply]
2022-05-02T20:39:19.0629460Z stubs/dbus_next/aio/message_bus.pyi:35: error: Signature of "release_name" incompatible with supertype "BaseMessageBus"
2022-05-02T20:39:19.0629924Z stubs/dbus_next/aio/message_bus.pyi:35: note:      Superclass:
2022-05-02T20:39:19.0630699Z stubs/dbus_next/aio/message_bus.pyi:35: note:          def release_name(self, name: str, callback: Optional[Callable[[Optional[ReleaseNameReply], Optional[Exception]], None]] = ...) -> Any
2022-05-02T20:39:19.0631224Z stubs/dbus_next/aio/message_bus.pyi:35: note:      Subclass:
2022-05-02T20:39:19.0631827Z stubs/dbus_next/aio/message_bus.pyi:35: note:          def release_name(self, name: str) -> Coroutine[Any, Any, ReleaseNameReply]
2022-05-02T20:39:19.0632481Z stubs/dbus_next/aio/message_bus.pyi:38: error: Argument 3 of "get_proxy_object" is incompatible with supertype "BaseMessageBus"; supertype defines the argument type as "Union[Node, str, Element]"
2022-05-02T20:39:19.0633083Z stubs/dbus_next/aio/message_bus.pyi:38: note: This violates the Liskov substitution principle
2022-05-02T20:39:19.0633798Z stubs/dbus_next/aio/message_bus.pyi:38: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
```

I generated stubs by doing `stubgen -p dbus_next -o stubs` but maybe it's quicker to do this manually for the bits we need.

Reply:
Maybe we don't need stubs here. `dubs-next` is a dependency of the `mypy` environment in tox which is how people should be running tests locally anyway.

Reply:
Just rethinking this... we only need stubs if the library's source isn't typed with a [py.typed](https://github.com/altdesktop/python-dbus-next/blob/master/dbus_next/py.typed) file. Without the library installed, `mypy` will emit a message saying that type information cannot be found and that the user should installed the library or get some stubs. In this case, the better approach is to just install the library because that's where the true type information is. So back to your question:

> What's the right approach in these scenarios?

mypy will handle telling the user what to do - they should install the library.